### PR TITLE
RFC: pp motives and misc delab fixes

### PR DESCRIPTION
**Pull request by dselsam**
This PR adds support for printing motives both in and out of matches. It also fixes two other minor issues in the delaborator that were needed to get `Std.Range.forIn.loop` to roundtrip. 

- Feature 1: `pp.pi_motives`. Print motives that return `pi` types, `true` by default.
- Feature 2: `pp.dep_motives`. Print motives whose return type depends on one of the arguments, `false` by default.
- Fix 1: delaborating lambdas was skipping instance arguments that the elaborator is expecting to see. Note: this fix solves a recurring problem I have been having, but there may still be other inconsistencies between the elaborator and delaborator regarding implicit lambas.
- Fix 2: one missing `:term` annotation inside a `doElem` was preventing the delaborated syntax tree from roundtripping. Note: this omission may have been intentional, since it could still roundtrip when indirecting through the parser, and adding the annotation introduced one additional set of parentheses in one of the tests.

Happy to split this up into 4 mini-commits and add additional tests/documentation.

### CODE REVIEW

File: src/Lean/PrettyPrinter/Delaborator/Basic.lean [hunk 1]
```diff
@@ -106,7 +106,16 @@ register_builtin_option pp.proofs.withType : Bool := {
   group    := "pp"
   descr    := "(pretty printer) when eliding a proof (see `pp.proofs`), show its type instead"
 }
-
+register_builtin_option pp.pi_motives : Bool := {
```
Comment (Kha) [inline src/Lean/PrettyPrinter/Delaborator/Basic.lean:109]: We switched to camelCase option names at some point, though we haven't updated all old names yet

File: src/Lean/PrettyPrinter/Delaborator/Builtins.lean [hunk 2]
```diff
@@ -128,6 +128,26 @@ def delabAppExplicit : Delab := do
...
+  | skip
+  | name (n : Name)
+
+def returnsPi (motive : Expr) : Bool := do
+  match motive with
+  | Expr.lam name d b _ => returnsPi b
+  | Expr.forallE .. => true
+  | _ => false
+
+def isDepFun (motive : Expr) : Bool := do
```
Comment (Kha) [inline src/Lean/PrettyPrinter/Delaborator/Builtins.lean:142]: The name is a bit misleading, it's more `isNonConstFun`...? Perhaps the use site could come with some examples. Also feel free to use `where` for helper functions that are probably not reusable.

File: src/Lean/PrettyPrinter/Delaborator/Builtins.lean [hunk 3]
```diff
@@ -137,15 +157,24 @@ def delabAppImplicit : Delab := whenNotPPOption getPPExplicit do
...
+      let opts ← getOptions
+      let argKind : ImplicitArgKind := match paramKinds with
+        | [ParamKind.implicit _ (some v)] => 
+          if !v.hasLooseBVars && v == arg then ImplicitArgKind.skip else ImplicitArgKind.provide
+        | ParamKind.implicit name none :: _  => do
+          if name == `motive && showMotive arg opts
+          then ImplicitArgKind.name name                        
+          else ImplicitArgKind.skip
+        | _ => ImplicitArgKind.provide
+      match argKind with
```
Comment (Kha) [inline src/Lean/PrettyPrinter/Delaborator/Builtins.lean:170]: I'd suggest to fuse this `match` with the prior `let`, getting rid of `ImplicitArgKind`, which doesn't seem to be used anywhere else. You can also factor out the common `(fnStx, paramKinds.tailD [],` part, then the `skip` and `provide` cases are trivial enough that duplicating them should not be an issue.

File: src/Lean/PrettyPrinter/Delaborator/Builtins.lean [hunk 4]
```diff
@@ -277,7 +309,13 @@ def delabAppMatch : Delab := whenPPOption getPPNotation do
   | _,        #[] => failure
   | _,        _   =>
     let pats ← delabPatterns st
-    let stx ← `(match $[$st.discrs:term],* with $[| $pats,* => $st.rhss]*)
+    let stx ← do
+      let (piStx, lamMotive) := st.motive.get!
+      let opts ← getOptions
+      if showMotive lamMotive opts then
+        `(match $[$st.discrs:term],* : $(piStx) with $[| $pats,* => $st.rhss]*)
```
Comment (Kha) [inline src/Lean/PrettyPrinter/Delaborator/Builtins.lean:316]:
```suggestion
        `(match $[$st.discrs:term],* : $piStx with $[| $pats,* => $st.rhss]*)
```
(unless I'm missing something)

File: tests/lean/run/pp_roundtrip_defs.lean [hunk 5]
```diff
@@ -0,0 +1,26 @@
+import Lean
```
Comment (Kha) [inline tests/lean/run/pp_roundtrip_defs.lean:1]: Please name test files in camelCase

File: src/Lean/PrettyPrinter/Delaborator/Builtins.lean [hunk 6]
```diff
@@ -606,7 +647,10 @@ partial def delabDoElems : DelabM (List Syntax) := do
       | Expr.lam _ _ body _ =>
         withBindingBodyUnusedName fun n => do
           if body.hasLooseBVars then
-            prependAndRec `(doElem|let $n:term ← $ma)
+            -- Note: the `$ma:term` annotation is required for elaboration to roundtrip
+            -- without indirecting through the parser. However, this annotation seems
+            -- to cause additional parentheses to be inserted in some cases.
```
Comment (Kha) [inline src/Lean/PrettyPrinter/Delaborator/Builtins.lean:652]: I only see *dropped* parentheses in the changed test output?

**Review by Kha — APPROVED** [review]

File: src/Lean/PrettyPrinter/Delaborator/Basic.lean (hunk 1 above)
Comment (dselsam) [reply src/Lean/PrettyPrinter/Delaborator/Basic.lean]: Fixed. Using these three now:
```lean
set_option pp.motives.pi true
set_option pp.motives.nonConst true
set_option pp.motives.all true
```

File: src/Lean/PrettyPrinter/Delaborator/Builtins.lean (hunk 2 above)
Comment (dselsam) [reply src/Lean/PrettyPrinter/Delaborator/Builtins.lean]: Yes, fixed.

File: src/Lean/PrettyPrinter/Delaborator/Builtins.lean (hunk 3 above)
Comment (dselsam) [reply src/Lean/PrettyPrinter/Delaborator/Builtins.lean]: I removed `ImplicitArgKind` at your suggestion. Now the `let` binds an `Option Syntax`, we match on this to compute the new `argStxs` and then all paths return together.

File: src/Lean/PrettyPrinter/Delaborator/Builtins.lean (hunk 4 above)
Comment (dselsam) [reply src/Lean/PrettyPrinter/Delaborator/Builtins.lean]: Fixed.

File: tests/lean/run/pp_roundtrip_defs.lean (hunk 5 above)
Comment (dselsam) [reply tests/lean/run/pp_roundtrip_defs.lean]: Fixed.

File: src/Lean/PrettyPrinter/Delaborator/Builtins.lean (hunk 6 above)
Comment (dselsam) [reply src/Lean/PrettyPrinter/Delaborator/Builtins.lean]: Yes, good. I removed this comment.

**Pull request merged**